remote_attestation/rust/src/message.rs

```diff
@@ -0,0 +1,251 @@
+    #[serde(with = "BigArray")]
+    pub signing_public_key: [u8; SIGNING_PUBLIC_KEY_SIZE_BYTES],
+    /// Information used for remote attestation such as a TEE report and a TEE provider's
+    /// certificate. TEE report contains a hash of the `signing_public_key`.
+    ///
+    /// Attestation info must be a serialized `oak.remote_attestation.AttestationInfo` Protobuf
+    /// message.
+    pub attestation_info: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct ClientIdentity {
+    /// Message header.
+    header: u8,
+    /// Public key needed to establish a session key.
+    pub ephemeral_public_key: [u8; EPHEMERAL_PUBLIC_KEY_SIZE_BYTES],
+    /// Signature of the SHA-256 hash of all previously send and received messaged.
+    /// Transcript signature is sent in messages to prevent replay attacks.
+    ///
+    /// Signature must be an IEEE-P1363 encoded ECDSA-P256 signature.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    /// https://standards.ieee.org/standard/1363-2000.html
+    #[serde(with = "BigArray")]
+    pub transcript_signature: [u8; TRANSCRIPT_SIGNATURE_SIZE_BYTES],
+    /// Public key used to sign transcripts.
+    ///
+    /// Public key must be an OpenSSL ECDSA-P256 key, which is represented as
+    /// `0x04 | X: 32-byte | Y: 32-byte`.
+    /// Where X and Y are big-endian coordinates of an Elliptic Curve point.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    #[serde(with = "BigArray")]
+    pub signing_public_key: [u8; SIGNING_PUBLIC_KEY_SIZE_BYTES],
+    /// Information used for remote attestation such as a TEE report and a TEE provider's
+    /// certificate. TEE report contains a hash of the `signing_public_key`.
+    ///
+    /// Attestation info must be a serialized `oak.remote_attestation.AttestationInfo` Protobuf
+    /// message.
+    pub attestation_info: Vec<u8>,
+}
+
+pub trait Serializable {
+    fn serialize(&self) -> anyhow::Result<Vec<u8>>;
+    fn deserialize(bytes: &[u8]) -> anyhow::Result<Self> where Self: Sized;
+}
+
+pub struct ClientHelloBuilder<'a> {
+    pub random: &'a [u8],
+}
+
+impl<'a> ClientHelloBuilder<'a> {
+    pub fn build(&'a self) -> anyhow::Result<ClientHello> {
+        Ok(ClientHello {
+            header: CLIENT_HELLO_HEADER,
+            random: to_array(self.random).context("Couldn't assign random vector")?,
+        })
+    }
+}
+
+impl Serializable for ClientHello {
+    fn serialize(&self) -> anyhow::Result<Vec<u8>> {
+        bincode::serialize(&self).context("Couldn't serialize client hello message")
+    }
+
+    fn deserialize(bytes: &[u8]) -> anyhow::Result<Self> {
+        let message: Self = bincode::deserialize(&bytes)
+            .context("Couldn't serialize client hello message")?;
+        if message.header == CLIENT_HELLO_HEADER {
+            Ok(message)
+        } else {
+            Err(anyhow!("Incorrect client hello message header"))
+        }
+    }
+}
+
+pub struct ServerIdentityBuilder<'a> {
```

Comment:
What is the benefit of having a builder? Most of the fields in `ServerIdentity` are public. The same for `ClientIdentity`.

Reply:
Because these fields are fixed size arrays `[u8; L]`. And creating such parameters will require and additional code each time.
Also, having a new method with `Vec<u8>`'s will be very hard to initialize without errors, because there will be no type checks.
So builder with named fields is just a readability addition.

Reply:
Why is additional code required for the arrays? As long as they are correctly typed from the beginning, the types should line up. I think the current code is turning a lot of checks that could be done at compile time into runtime checks.

Reply:
Done